genetic_algorithms/problems/graph_coloring_problem/visualization.py

```diff
@@ -0,0 +1,145 @@
+from pathlib import Path
+from math import inf
+
+import pygame
+import sys
+import time
+import random
+
+from genetic_algorithms.algorithm_wrappers.algorithm_wrapper import AlgorithmWrapper
+from genetic_algorithms.models.next_state_provider import NextStateProvider
+from genetic_algorithms.problems.graph_coloring_problem.problem_model import GraphColoringModel
+from genetic_algorithms.problems.graph_coloring_problem.state import GraphColoringState
+from genetic_algorithms.solvers.solver import SolverConfig
+
+class GraphColoringVisualization(AlgorithmWrapper):
+    def __init__(self, config: SolverConfig, algorithm: NextStateProvider, model: GraphColoringModel, **kwargs):
+        pygame.init()
+        pygame.font.init()
+        self.time_limit = config.time_limit
+        self.coord = self._generate_coord(model)
+        self.start_time = time.time()
+        self.time_limit = config.time_limit
+        self.colors = self._generate_colors(model)
+        self.states = 0
+        self.cost_best_state = inf
+        super().__init__(algorithm, **kwargs)
+
+    def _perform_side_effects(self, model: GraphColoringModel, state: GraphColoringState):
+        self.screen = pygame.display.set_mode((800, 800))
+        self.current_time = time.time()
+        self.states += 1
+        self.cost_current_state = model.cost_for(state)
+        self._check_states()
+
+        while True:
+            self._handle_pygame_events()
+            self._draw(state, model)
+
+    def _draw_vertices(self, state: GraphColoringState):
+        for i in range(len(self.coord)):
+            x, y = self._scale(self.coord[i])
+            pygame.draw.circle(self.screen, self.colors[state.coloring[i].color], (x, y), 10)
+            pygame.draw.circle(self.screen, (0, 0, 0), (x, y), 10, 2)
+
+    def _draw(self, state: GraphColoringState, model: GraphColoringModel):
+        self.screen.fill((255, 255, 255))
+        self._draw_lines(model)
+        self._draw_vertices(state)
+        self._draw_information()
+        pygame.display.flip()
+
+        if self._check_time():
+            while True:
+                self.screen = pygame.display.set_mode((900, 900))
+                self._handle_pygame_events()
+
+    def _generate_coord(self, model: GraphColoringModel):
+        coord = []
+        x, y = 1, 1
+        for i in range(model.n_vertices):
+            while (x, y) in coord:
+                x, y = random.randint(0, model.n_vertices), random.randint(0, model.n_vertices)
+            coord.append((x, y))
+
+        return coord
+
+    def _generate_colors(self, model: GraphColoringModel):
+        colors = []
+        color = (0, 150, 0)
+        for i in range(model.n_vertices):
+            while color in colors:
+                color = (random.randint(0, 255), random.randint(0, 255), random.randint(0, 255))
+            colors.append(color)
```

Comment:
Może się tak trafić, że wygeneruje bardzo podobne kolory. Można to jakoś wyeliminować?